core/src/main/java/org/apache/calcite/plan/RelOptPredicateList.java

```diff
@@ -104,6 +106,36 @@ private RelOptPredicateList(ImmutableList<RexNode> pulledUpPredicates,
     this.rightInferredPredicates =
         requireNonNull(rightInferredPredicates, "rightInferredPredicates");
     this.constantMap = requireNonNull(constantMap, "constantMap");
+
+    // Validate invariants required
+    // (unfortunately the style rules don't allow us to move this to a separate function).
+    // Do not allow comparisons with null literals in pulledUpPredicates
+    for (RexNode predicate : this.pulledUpPredicates) {
```

Comment:
The constructor of the current class is private. If this class can only be created through the `of` method, can this code be moved to the `of` method and encapsulated into a new method?

Reply:
Yes, but it someone creates a new constructor, it may get overlooked.

Reply:
If it helps, you could add a comment that the private constructor must remain private, and must remain the only constructor.